src/ParseServer.js

```diff
@@ -82,38 +82,40 @@ class ParseServer {
 
     logging.setLogger(loggerController);
 
+    if (cloud) {
+      addParseCloud();
+      if (typeof cloud === 'function') {
+        cloud(Parse);
+      } else if (typeof cloud === 'string') {
+        require(path.resolve(process.cwd(), cloud));
+      } else {
+        throw "argument 'cloud' must either be a string or a function";
+      }
+    }
+
     // Note: Tests will start to fail if any validation happens after this is called.
-    databaseController
-      .performInitialization()
-      .then(() => hooksController.load())
+    Promise.resolve()
+      .then(async () => await databaseController.performInitialization())
+      .then(async () => await hooksController.load())
       .then(async () => {
         if (schema) {
           await new DefinedSchemas(schema, this.config).execute();
         }
+      })
+      .then(async () => {
         if (serverStartComplete) {
-          serverStartComplete();
+          await serverStartComplete();
         }
       })
-      .catch(error => {
+      .catch(async (error) => {
         if (serverStartComplete) {
-          serverStartComplete(error);
+          await serverStartComplete(error);
         } else {
           console.error(error);
           process.exit(1);
         }
       });
```

Comment:
Just to help to cross de finish line, here the full correct structure.

```js
    // Note: Tests will start to fail if any validation happens after this is called.
    databaseController.performInitialization()
      .then(() => hooksController.load())
      .then(async () => {
        if (schema) {
          await new DefinedSchemas(schema, this.config).execute();
        }
      })
      .then(async () => {
        if (serverStartComplete) {
          await Promise.resolve(serverStartComplete());
        }
      })
      .catch(async (error) => {
        if (serverStartComplete) {
          await Promise.resolve(serverStartComplete(error));
        } else {
          console.error(error);
          process.exit(1);
        }
      });
```

Then in a futur PR that refactor the start up (BK change) the structure will be 

```js
try {
  await databaseController.performInitialization()
  await hooksController.load()
  if (schema) {
    await new DefinedSchemas(schema, this.config).execute();
  }
  if (serverStartComplete) {
    await Promise.resolve(serverStartComplete());
  }
} catch(e){
  if (serverStartComplete) {
      await Promise.resolve(serverStartComplete(error));
  } else {
    console.error(error);
    process.exit(1);
  }
}
```

Reply:
Mmm, not sure what's the difference.

1)
I don't remember where, but we had a discussion about starting Promises chains with `Promise.resolve()` to be consistent with other promises chains in Server and JS SDK


2) `await Promise.resolve(serverStartComplete());` 

I don't think we need to wrap it around `Promise.resolve`, `await` will work either way.